![Clique](https://github.com/mckennalab/clique/raw/main/clique_logo.png)

Clique aims to be a high-performance, unified package developed for the efficient processing and analysis of lineage data from Illumina and long-read technologies like Nanopore. It effectively collapses diverse unique molecular identifiers (UMIs) and static identifiers (SIs) and generates consensus sequences. The package aligns these consensus sequences to amplicon sequences, matching them to the best candidate sequence through a variety of alignment strategies, such as affine gap, convex alignments, and Hidden Markov Models (HMMs).

Clique supports a broad range of single-cell sequencing technologies, and it provides comprehensive statistics on read recovery, UMI complexity, alignment quality, and single-cell alignment. The package also enables the creation of lineage trees and statistics, empowering researchers with a deeper understanding of the genetic relationships between individual cells.
